### PR TITLE
Tbl.add instead of Tbl.replace for RO instances

### PR DESCRIPTION
**Pull request by icristescu**
We can use `Tbl.add` instead of `Tbl.replace` when we are adding values from a log or index (where there are no duplicate values) into a clear table. 

In the RO instance (in `sync_log`) we can always use `Tbl.add`.

### CODE REVIEW

File: src/index.ml [hunk 1]
```diff
@@ -345,7 +345,7 @@ struct
             l "[%s] log file detected. Loading %Ld entries"
               (Filename.basename root) entries);
         let mem = Tbl.create (Int64.to_int entries) in
-        iter_io (fun e -> Tbl.replace mem e.key e.value) io;
+        iter_io (fun e -> Tbl.add mem e.key e.value) io;
```
Comment (pascutto) [inline src/index.ml:348]: This is RW code so it will ultimately lead to duplicates in the index file.

Comment (icristescu) [reply src/index.ml:348]: But it is loading entries from the log file, which shouldn't contain duplicates, no?

Comment (icristescu) [reply src/index.ml]: Yes, you are right. Thanks !

I changed it back and also added two tests for duplicate entries, (one that fails if add is used here). In the test I noticed that `iter` calls iter on index twice, so I remove one of them.

File: src/index.ml [hunk 2]
```diff
@@ -753,7 +755,6 @@ struct
     | None -> ()
     | Some log ->
         Tbl.iter f log.mem;
-        may (fun (i : index) -> iter_io (fun e -> f e.key e.value) i.io) t.index;
```
Comment (samoht) [inline src/index.ml:756]: why did you remove this line?

Comment (icristescu) [reply src/index.ml:756]: the iter on index is called again a bit lower (https://github.com/mirage/index/blob/master/src/index.ml#L761), but git diff is not showing it.

Comment (samoht) [reply src/index.ml:756]: yes but why is this necessary to remove that line? Is that related to the change between replace and add?

Comment (icristescu) [reply src/index.ml:756]: I think these tests are related to this PR, but I can move both the tests and the change to `iter` in a different PR, if you think is clearer.

Comment (samoht) [reply src/index.ml:756]: well that's a bit suspicious if we have to change the semantics of a function to make the tests pass :-) @pascutto do you remember why `iter` is iterating over the elements twice? That sounds like a bug?

Comment (pascutto) [reply src/index.ml:756]: I'll check what's going on here.
A side note though: `iter` may hit some keys twice even if there are in fact no duplicates so it's not super robust to write tests assuming that `iter` reflects the correct state of duplication in the structure. (see our docstring here: https://github.com/mirage/index/blob/master/src/index.mli#L176-L183)

**Comment by samoht** [comment]
I still interested to run benchmarks with that PR if you case you have a bit of time to rebase it :-)

**Comment by icristescu** [comment]
I rebased this, but it only replaces `Tbl.replace` for RO instances (updated the title of the PR to reflect that), so it's probably not going to change much in the layers benchmark.

**Comment by samoht** [comment]
Closing this one as it only applies to read-only instance (and it's not clear it super useful).
